### PR TITLE
Update jvmtiGetThreadGroupChildren for java 19

### PR DESCRIPTION
**Pull request by EricYangIBM**
The ThreadGroup class does not store its threads anymore and child groups
are separated into "groups" and "weaks" ThreadGroup arrays.

The child ThreadGroups array is populated from the ThreadGroup's "groups"
and "weaks" arrays, which are the child ThreadGroups that are strongly
and weakly reachable from the parent.

The threads array is constructed by iterating over all J9VMThreads and
adding the threads that belong in the given ThreadGroup.

The returned arrays are constructed while locking on the ThreadGroup
object.

Fixes: #15244
Issue: #15183
Signed-off-by: Eric Yang <eric.yang@ibm.com>

### CODE REVIEW

**Review by babsingh — CHANGES_REQUESTED** [review]
The commit message/description should include few lines on how the `threads` and `groups` arrays are created for Java 19+. It will make it easier to follow the Java19+ implementation.

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (EricYangIBM) [reply runtime/jvmti/jvmtiThreadGroup.c:407]: We return the actual number of threads in the array so it probably is fine

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (EricYangIBM) [reply runtime/jvmti/jvmtiThreadGroup.c:26]: In other files the `static jvmtiError` and function name are on the same line. Should I follow their formatting?

Comment (babsingh) [reply runtime/jvmti/jvmtiThreadGroup.c:26]: Match existing formatting. Also, you can keep multiple parameters on a single line if needed. It should be consistent.

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (babsingh) [reply runtime/jvmti/jvmtiThreadGroup.c:420]: A lot of allocated memory will be unused if the filter excludes a lot of threads. Filter: include only platform threads which match the thread_group.

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (babsingh) [reply runtime/jvmti/jvmtiThreadGroup.c:422]: A lot of allocated memory will be unused if a lot of weak references are null.

**Comment by babsingh** [comment]
@EricYangIBM The commits can be squashed.

@tajila These changes should work fine. But, still doubtful about the memory allocated for `threads` and `groups` arrays which are returned from this method (see comments inlined in code). Should we create new arrays with smaller/required size and copy the elements from the bigger array and free the bigger array?

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (EricYangIBM) [reply runtime/jvmti/jvmtiThreadGroup.c:404]: According to your comment https://github.com/eclipse-openj9/openj9/issues/15183#issuecomment-1167767823
the J9VMThread link list does not include virtual threads.

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (EricYangIBM) [inline runtime/jvmti/jvmtiThreadGroup.c:400]: To do: change to targetThread->carrierThreadObject

**Comment by tajila** [comment]
> But, still doubtful about the memory allocated for threads and groups arrays which are returned from this method (see comments inlined in code). Should we create new arrays with smaller/required size and copy the elements from the bigger array and free the bigger array?

This is not going to return vthreads correct? If so, then I dont think returning a potentially larger buffer is a big deal as platform threads arent created as frequently

**Comment by tajila** [comment]
It could be an optimization that is added later

**Comment by babsingh** [comment]
> This is not going to return vthreads correct?

Vthreads will not be returned. See ref below. @EricYangIBM Can you add a comment in code and open an issue regarding this optimization for tracking purposes?

Ref: https://download.java.net/java/early_access/jdk19/docs/specs/jvmti.html#GetThreadGroupChildren

File: runtime/jvmti/jvmtiThreadGroup.c
Comment (EricYangIBM) [reply runtime/jvmti/jvmtiThreadGroup.c]: Done

**Comment by EricYangIBM** [comment]
I will open the issue after this PR is merged (to link to the comment)

**Comment by babsingh** [comment]
LGTM. The commits need to be squashed before the PR builds are launched.

**Comment by EricYangIBM** [comment]
Squashed

**Comment by babsingh** [comment]
jenkins compile win jdk19,jdk17

**Comment by babsingh** [comment]
Does not work for me yet. Need to wait for @AdamBrousseau to add me.

**Comment by babsingh** [comment]
jenkins compile win jdk19,jdk17

**Comment by babsingh** [comment]
These JVMTI changes should only impact the `gtgc001` test, which has been enabled and passes for JDK17 and JDK19. 

> https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.functional_aarch64_mac_Personal/75/

`jit_vich_0_FAILED` is unrelated to this PR. All tests in this suite pass but there is an error at the end (`Invalid JIT return address`).
```
===============================================
JIT_Test suite
Total tests run: 18, Failures: 0, Skips: 0
===============================================

*** Invalid JIT return address 000000037A4175E0 in 000000016FBDFD98

jit_vich_0_FAILED
```

**Review by babsingh — APPROVED** [review]
LGTM. Opened https://github.com/eclipse-openj9/openj9/issues/15673 for the JIT failure.

**Comment by babsingh** [comment]
Still can't merge the PR. Need to wait for @AdamBrousseau to give the required permissions.

**Comment by AdamBrousseau** [comment]
>Still can't merge the PR

This is an Eclipse Github team thing. It should be automatic. If it's not working then I would open an Eclipse ticket.
https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues?sort=created_date&state=opened

**Pull request merged**
